File: aircon/src/main/java/com/ironsource/aura/aircon/AirCon.java
package com.ironsource.aura.aircon;

import android.app.Application;
import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.ironsource.aura.aircon.injection.AttributeResolver;
import com.ironsource.aura.aircon.logging.Logger;
import com.ironsource.aura.aircon.source.ConfigSource;
import com.ironsource.aura.aircon.source.ConfigSourceRepository;

/**
 * AirCon SDK entry point.
 * Use {@link #get()} in order to get an instance.
 */
public class AirCon {

	private static AirCon sInstance;

	private Context mContext;

	private Logger            mLogger;
	private Class             mAttrClass;
	private AttributeResolver mAttributeResolver;
	private JsonConverter     mJsonConverter;

	private ConfigSourceRepository mConfigSourceRepository;

	/**
	 * Returns a singleton instance of the AirCon SDK.
	 * Before any use of the SDK {@link #init(AirConConfiguration)} should be called ({@link Application#onCreate()} is the common place)
	 *
	 * @return AirCon SDK instance.
	 */
	public static synchronized AirCon get() {
		if (sInstance == null) {
			synchronized (AirCon.class) {
				if (sInstance == null) {
					sInstance = new AirCon();
				}
			}
		}
		return sInstance;
	}

	/**
	 * Initializes the SDK with the provided configuration.
	 *
	 * @param airConConfiguration SDK configuration.
	 */
	public void init(@NonNull AirConConfiguration airConConfiguration) {
		if (isInitialized()) {
			return;
		}
		mContext = airConConfiguration.getContext();

		mLogger = airConConfiguration.getLogger();
		mAttrClass = airConConfiguration.getAttrClass();
		mAttributeResolver = airConConfiguration.getAttributeResolver();
		mJsonConverter = airConConfiguration.getJsonConverter();

		final SdkContext sdkContext = new SdkContext(airConConfiguration.getLogger());
		mConfigSourceRepository = new ConfigSourceRepository(sdkContext, airConConfiguration.getConfigSources(), airConConfiguration.getIdentifiableConfigSources(), airConConfiguration.getIdentifiableConfigSourceFactories());
	}

	/**
	 * Returns whether the SDK was initialized.
	 *
	 * @return true is SDK is initialized, false otherwise.
	 * @see #init(AirConConfiguration)
	 */
	public boolean isInitialized() {
		return mContext != null;
	}

	/**
	 * Returns the application context retrieved from context the SDK was initialized with.
	 *
	 * @return SDK context.
	 * @throws IllegalStateException if SDK is not initialized.
	 */
	@NonNull
	public Context getContext() {
		assertInitialized();
		return mContext;
	}

	/**
	 * Returns the SDK logger set by {@link AirConConfiguration.Builder#setLogger(Logger)}
	 * or {@link com.ironsource.aura.aircon.logging.AndroidLogger} if non is provided.
	 *
	 * @return SDK logger.
	 * @throws IllegalStateException if SDK is not initialized.
	 */
	@NonNull
	public Logger getLogger() {
		assertInitialized();
		return mLogger;
	}

	/**
	 * Returns whether XML injection is enabled.
	 *
	 * @return true if XML injection is enabled, false otherwise.
	 * @throws IllegalStateException if SDK is not initialized.
	 * @see AirConConfiguration.Builder#enableXmlInjection(Class, AttributeResolver)
	 */
	public boolean isXmlInjectionEnabled() {
		assertInitialized();
		return mAttrClass != null;
	}

	/**
	 * Returns the attribute class if supplied, null otherwise.
	 *
	 * @return attribute class if supplied, null otherwise.
	 * @throws IllegalStateException if SDK is not initialized.
	 * @see AirConConfiguration.Builder#enableXmlInjection(Class, AttributeResolver)
	 */
	@Nullable
	public Class getAttrClass() {
		assertInitialized();
		return mAttrClass;
	}

	/**
	 * Returns attribute resolver if supplied, null otherwise.
	 *
	 * @return attribute resolver if supplied, null otherwise.
	 * @throws IllegalStateException if SDK is not initialized.
	 * @see AirConConfiguration.Builder#enableXmlInjection(Class, AttributeResolver)
	 */
	public AttributeResolver getAttributeResolver() {
		assertInitialized();
		return mAttributeResolver;
	}

	/**
	 * Returns json converter if supplied..
	 *
	 * @return json converter if supplied, null otherwise.
	 * @throws IllegalStateException if SDK is not initialized or no converter is supplied.
	 * @see AirConConfiguration.Builder#setJsonConverter(JsonConverter)
	 */
	public JsonConverter getJsonConverter() {
		assertInitialized();

		if (mJsonConverter == null) {
			throw new IllegalStateException("No json converter available, a converter needs to be supplied in the AirConConfiguration");
		}

		return mJsonConverter;
	}

	/**
	 * Returns config sources repository through which {@link ConfigSource}
	 * can be added, removed and retrieved.
	 *
	 * @return config source repository.
	 * @throws IllegalStateException if SDK is not initialized.
	 */
	@NonNull
	public ConfigSourceRepository getConfigSourceRepository() {
		assertInitialized();
		return mConfigSourceRepository;
	}

	private void assertInitialized() {
		if (!isInitialized()) {
			throw new IllegalStateException("Sdk not initialized, did you call AirCon.init()?");
		}
	}
}

File: aircon/src/main/java/com/ironsource/aura/aircon/AirCon.kt
package com.ironsource.aura.aircon

import android.content.Context
import com.ironsource.aura.aircon.AirCon.init
import com.ironsource.aura.aircon.common.ConfigTypeResolver
import com.ironsource.aura.aircon.injection.AttributeResolver
import com.ironsource.aura.aircon.logging.Logger
import com.ironsource.aura.aircon.source.ConfigSourceRepository
import kotlin.properties.ReadWriteProperty
import kotlin.reflect.KProperty

/**
 * AirCon SDK entry point.
 * Before any use of the SDK [AirCon][init] should be called (Application.onCreate() is the common place)
 */
@Suppress("MemberVisibilityCanBePrivate")
object AirCon {

    /**
     * Whether the SDK was initialized.
     *
     * @see init
     */
    var initialized: Boolean = false
        private set

    /**
     * The application context retrieved from context the SDK was initialized with.
     *
     * @throws IllegalStateException if SDK is not initialized.
     */
    var context: Context? by InitializedBarrierDelegate()
        private set

    /**
     * The SDK logger set by [AirConConfiguration.Builder.setLogger]
     * or [com.ironsource.aura.aircon.logging.AndroidLogger] if non is provided.
     *
     * @throws IllegalStateException if SDK is not initialized.
     */
    var logger: Logger? by InitializedBarrierDelegate()
        private set

    /**
     * Attribute class if supplied
     *
     * @throws IllegalStateException if SDK is not initialized.
     * @see AirConConfiguration.Builder.enableXmlInjection
     */
    var attrClass: Class<*>? by InitializedBarrierDelegate()
        private set

    /**
     * Attribute resolver if supplied
     *
     * @throws IllegalStateException if SDK is not initialized.
     * @see AirConConfiguration.Builder.enableXmlInjection
     */
    var attributeResolver: AttributeResolver? by InitializedBarrierDelegate()
        private set

    /**
     * Returns json converter if supplied.
     *
     * @throws IllegalStateException if SDK is not initialized or no converter is supplied.
     * @see AirConConfiguration.Builder.setJsonConverter
     */
    var jsonConverter: JsonConverter? by InitializedBarrierDelegate {
        check(it != null) { "No json converter available, a converter needs to be supplied in the AirConConfiguration" }
    }
        private set

    /**
     * Config sources repository through which [com.ironsource.aura.aircon.common.ConfigSource]
     * can be added, removed and retrieved.
     *
     * @throws IllegalStateException if SDK is not initialized.
     */
    var configSourceRepository: ConfigSourceRepository? by InitializedBarrierDelegate()
        private set

    private lateinit var mConfigTypes: Map<Class<out Annotation>, ConfigTypeResolver<*, *, *>>

    /**
     * Initializes the SDK with the provided configuration.
     *
     * @param airConConfiguration SDK configuration.
     */
    fun init(airConConfiguration: AirConConfiguration) {
        if (initialized) {
            return
        }

        initialized = true

        context = airConConfiguration.context

        logger = airConConfiguration.logger
        attrClass = airConConfiguration.attrClass
        attributeResolver = airConConfiguration.attributeResolver
        jsonConverter = airConConfiguration.jsonConverter
        mConfigTypes = airConConfiguration.configTypes

        val sdkContext = SdkContext(airConConfiguration.logger)
        configSourceRepository = ConfigSourceRepository(sdkContext, airConConfiguration.configSources, airConConfiguration.identifiableConfigSources, airConConfiguration.identifiableConfigSourceFactories)
    }

    /**
     * Returns whether XML injection is enabled.
     *
     * @throws IllegalStateException if SDK is not initialized.
     * @see AirConConfiguration.Builder.enableXmlInjection
     */
    fun isXmlInjectionEnabled(): Boolean {
        assertInitialized()

        return attrClass != null
    }

    @Suppress("UNCHECKED_CAST")
    fun <A : Annotation, T, S> getConfigTypeResolver(configTypeAnnotation: Class<A>): ConfigTypeResolver<A, T, S> {
        assertInitialized()

        val configTypeResolver = mConfigTypes[configTypeAnnotation] as ConfigTypeResolver<A, T, S>?

        checkNotNull(configTypeResolver) {
            "No config resolver found for ${configTypeAnnotation.simpleName}, custom config types should be registered in the AirConConfiguration"
        }

        return configTypeResolver
    }
}

private class InitializedBarrierDelegate<T>(private val extCheck: ((T?) -> Unit)? = null) : ReadWriteProperty<AirCon, T?> {
    var value: T? = null

    override fun getValue(thisRef: AirCon, property: KProperty<*>): T? {
        assertInitialized()

        extCheck?.invoke(value)

        return value
    }

    override fun setValue(thisRef: AirCon, property: KProperty<*>, value: T?) {
        this.value = value
    }
}

private fun assertInitialized() {
    check(AirCon.initialized) { "Sdk not initialized, did you call AirCon.init()?" }
}
